### PR TITLE
Automatically Generate a Config File Based on Samplesheet

### PR DESCRIPTION
**Pull request by emarinier**
This pull request adds the ability to automatically generate an [Arborator config JSON file](https://github.com/phac-nml/arborator/blob/master/arborator/test_data/config.json) if one is not provided. This is important because Arborator cannot be run without such a file, and it may prove difficult to expect users to generate this file on their own.

Generation of the config file is achieved using a lookup file ([assets/config_lookup.json](assets/config_lookup.json)) which has the expected formatting of an Arborator config file (such that the lookup could be used as a config file, but shouldn't be). This lookup file has many entries for expected typical inputs. The workflow uses information in the samplesheet (particularly the headers) to automatically build a config file. We should continue to add more typical entries to the lookup file.

If a config file is provided by the user (such as in the case of a very standardized workflow), the workflow will use the user-provided config file and won't generate another config file.

Additionally, some changes have been made to automatically detect and remove metadata columns in the samplesheet which contain no data, thereby producing cleaner Arborator output. Some vestigial example pipeline code has also been removed.

### CODE REVIEW

File: assets/config_lookup.json
Comment (mattheww95) [inline assets/config_lookup.json:1]: For `clustering_threshold` is it needed or is there a need to have some way of checking if the distances used are the scaled ones? If so do the default clustering thresholds need to change to reflect this?

**Review by mattheww95 — APPROVED** [review]

File: tests/modules/local/maptotsv/main.nf.test [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        then {
+            assert process.success
+
+            with(process.out) {
+                // check outputs
+                assert path(tsv_path[0]).exists()
+                assert nonempty_column_headers == [["sample", "outbreak", "organism", "subtype", "country", "serovar", "age", "date", "source", "special"]]
+
+                // parse tsv file
+                def tsv_text = path(tsv_path[0]).readLines()
```
Comment (apetkau) [inline tests/modules/local/maptotsv/main.nf.test:30]: Can you add two additional assert statements for each test:

1. Assert that the appropriate header lines are contained in the output file.
2. Assert that the number of lines is correct (in this case, should be 2 lines).

File: modules/local/map_to_tsv.nf [hunk 2]
```diff
@@ -23,11 +24,21 @@ process MAP_TO_TSV {
     def delimiter = '\t'
     output_place = task.workDir.resolve(output_file)
 
-    output_place.withWriter{ writer ->
+    metadata = [metadata_headers] + metadata_rows
+    metadata = metadata.transpose()
+    
+    for (int i in (metadata.size()-1)..0) { // Reverse order
+        if (metadata[i][1..-1].every {it == ""}) { // Every except for header
```
Comment (apetkau) [inline modules/local/map_to_tsv.nf:31]: Could you include a check for null strings in addition to empty strings (e.g, `it == null || it == ""`)

File: tests/modules/local/maptotsv/main.nf.test [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            process {
+                """
+                input[0] = new Tuple("sample", "outbreak", "organism", "subtype", "country", "serovar", "age", "date", "source", "special")
+                input[1] = new Tuple(["S1", "1", "Escherichia coli", "EHEC/STEC", "Canada", "O157:H7", "", "", "", ""])
+                """
+            }
+        }
+
+        then {
+            assert process.success
```
Comment (apetkau) [inline tests/modules/local/maptotsv/main.nf.test:82]: Could you add additional asserts here on the contents of the output file like the other test cases.

File: tests/modules/local/maptotsv/main.nf.test [hunk 4]
```diff
@@ -0,0 +1,289 @@
+nextflow_process {
+
+    name "Test Process MAP_TO_TSV"
```
Comment (apetkau) [inline tests/modules/local/maptotsv/main.nf.test:3]: Could you add additional test cases to this file/make sure they are handled in the code:

1. Instead of empty string `""` as missing values could you test the case of `null` as missing values (in case `null` ever is returned by Nextflow for one of the metadata values).
2. Could you test a case of one or more headers being `null` or empty (in case one of the header names was set to `null`).
3. Could you test the case where the number of columns for headers and rows are mismatched (e.g., 2 headers, but 1 value in a row)? I would expect the process to raise an exception/return an error in this case. I know this won't happen with your current code, but it's good to have a check in case refactoring down the line led to this situation.
4. Could you test the case of passing all empty data (empty headers, empty rows).
5. Could you test the case of the function returning all empty data (e.g., all columns are removed)?

File: modules/local/buildconfig/main.nf [hunk 5]
```diff
@@ -0,0 +1,93 @@
+import groovy.json.JsonSlurper
+import groovy.json.JsonOutput
+
+process BUILD_CONFIG {
+    tag "Build Arborator Config"
+    label 'process_single'
+
+    input:
+    val metadata_headers
+    // tuple(ID (header name), metadata_partition_name, metadata_1_header, .. , metadata_8_header)
```
Comment (apetkau) [inline modules/local/buildconfig/main.nf:10]: Could you either move this on the same line as `metadata_headers` or above that line? I thought at first this was another input parameter that was commented out

File: tests/pipelines/main.nf.test [hunk 6]
```diff
@@ -37,6 +35,22 @@ nextflow_pipeline {
             def expected_metadata = path("$baseDir/tests/data/metadata/expected_merged_data.tsv")
             assert actual_metadata.text == expected_metadata.text
 
+            // Check auto-built config file:
+            def actual_config = path("$launchDir/results/build/config.json")
+            def expected_config = path("$baseDir/tests/data/configs/autoconfig_samplesheet.json")
+            assert actual_config.text == expected_config.text
+
+            // Arborator outputs:
+            // NOTE: Arborator produces a (somewhat) non-deterministic cluster summary,
```
Comment (apetkau) [inline tests/pipelines/main.nf.test:44]: In what way is it non-deterministic? Is it related to the order samples are processed in channels? Since we could enforce a certain order in order to make deterministic in that case.

File: assets/config_lookup.json
Comment (apetkau) [reply assets/config_lookup.json:1]: As a note for later, I'm not quite sure what the difference between `clustering_threshold` in this JSON file is, and the `params.ar_thresholds` passed to the pipeline (and to arborator eventually). I am wondering if we could set the config `clustering_threshold` from a parameter to the pipeline like this?

**Review by apetkau — CHANGES_REQUESTED** [review]
This looks amazing @emarinier . Thanks so much for all your work 😄 

This is an initial review of the PR, but I may have other comments later. Thanks again.

**Comment by kylacochrane** [comment]
This is amazing Eric! Such a great solution to auto-generate the Arborator config JSON file!

**Review by kylacochrane — APPROVED** [review]

File: tests/modules/local/maptotsv/main.nf.test (hunk 3 above)
Comment (emarinier) [reply tests/modules/local/maptotsv/main.nf.test:82]: Yeah, this is weird. I wonder why I left it empty. Haha.

File: tests/modules/local/maptotsv/main.nf.test (hunk 1 above)
Comment (emarinier) [reply tests/modules/local/maptotsv/main.nf.test:30]: Added in 0ebd7271c224d509091af468c63bf5313067bcb5

File: tests/modules/local/maptotsv/main.nf.test (hunk 3 above)
Comment (emarinier) [reply tests/modules/local/maptotsv/main.nf.test:86]: Added in 0ebd7271c224d509091af468c63bf5313067bcb5

File: modules/local/map_to_tsv.nf (hunk 2 above)
Comment (emarinier) [reply modules/local/map_to_tsv.nf]: Added in 5695b445a1a8fdba6248e855f593cf73abd99e03

File: tests/pipelines/main.nf.test (hunk 6 above)
Comment (emarinier) [reply tests/pipelines/main.nf.test:44]: So, what happens is that I can specify in the arborator config file the order of some columns, but then arborator will add a bunch of summary columns (automatically and unspecified) that are derived from the "main" columns.

For example, if I build an arborator config file that has the following `grouped_metadata_columns`: `outbreak`, `organism`, `subtype`, `country`, `serovar`, `age`, `date`, `source`, `special`, then I'll get an `arborator/cluster_summary.tsv` file that has these headers:

```
Outbreak Code   Organism        Subtype Country of Collection   Serovar Patient Age (years)     Date    Source Type     Special age_mean_value  count_country_The United States   count_serovar_O125      count_subtype_EAEC      date_min_value  count_source_cheese     count_source_fruit      max_dist        count_age_55    count_age_61      count_members   count_serovar_O111:H21  count_subtype_EPEC      count_age_21    count_age_35    age_min_value   min_dist        count_outliers  count_serovar_O126:H27    count_country_France    count_special_False     count_organism_Escherichia coli count_special_True      count_age_14    count_source_beef         count_country_Canada    date_max_value  count_serovar_O157:H7   count_subtype_EHEC/STEC count_age_43    count_source_milk       median_dist     age_max_value     age_median_value        mean_dist
```

However, if I run the workflow again, I'll get:

```
Outbreak Code   Organism        Subtype Country of Collection   Serovar Patient Age (years)     Date    Source Type     Special count_special_True      count_country_The United States   age_min_value   mean_dist       median_dist     count_subtype_EAEC      count_source_beef       count_source_milk       age_max_value   count_organism_Escherichia coli   count_serovar_O125      count_country_France    count_serovar_O157:H7   count_country_Canada    max_dist        count_age_55    count_source_cheese       count_source_fruit      count_subtype_EHEC/STEC age_mean_value  age_median_value        count_members   count_age_14    date_min_value  count_serovar_O111:H21    count_special_False     count_age_21    count_age_43    count_age_35    count_serovar_O126:H27  count_outliers  count_subtype_EPEC      min_dist  date_max_value  count_age_61
```

Basically, everything after my specifications (which are mapped from keys to labels: `outbreak` -> `Outbreak Code`), which means `Special` onwards, are effectively non-deterministic. I can't guarantee the ordering of these columns, so I'm only testing the part that is deterministic.

File: modules/local/buildconfig/main.nf (hunk 5 above)
Comment (emarinier) [reply modules/local/buildconfig/main.nf]: Sure thing: f70f439436e0fcb6816e28c31871f0c084e53b91

File: tests/pipelines/main.nf.test (hunk 6 above)
Comment (apetkau) [reply tests/pipelines/main.nf.test:44]: Okay. Thanks so much for the explanation. There doesn't seem to be an easy way to make the output deterministic so what you are doing makes sense.

File: tests/modules/local/maptotsv/main.nf.test (hunk 4 above)
Comment (emarinier) [reply tests/modules/local/maptotsv/main.nf.test:3]: Added in 08ee7bd9b5de0e0b3f72a9b4320ab902efd7be16 and 1dd07cbec1d626db4c8cc0e63810cc5110c56ba1

Note that item 4 causes item 5, so they're tested at the same time.

File: assets/config_lookup.json
Comment (emarinier) [reply assets/config_lookup.json:1]: I've decided to remove `clustering_threshold` and `clustering_method`, which appears to have had no effect on the test output: 24ba00c38be88432fb05bd737d7f5ebefbccdb12

**Review by apetkau — APPROVED** [review]
Thanks so much Eric. This looks great 😄
